core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/cache/ScenarioCache.java

```diff
@@ -0,0 +1,554 @@
+      if (coveredInstanceRole != null && coveredInstanceRole.equals(instanceRole)) {
+        TimeLapse timeLapse = getTimeLapseFromCache(end);
+
+        // Execution End case
+        if (end instanceof ExecutionEnd) {
+          visit(ancestors, timeLapse, (ExecutionEnd) end, result);
+        }
+
+        // Interaction State case
+        if (end instanceof InteractionState) {
+          visit(ancestors, timeLapse, (InteractionState) end, result);
+        }
+
+        // Message End case
+        if (end instanceof MessageEnd) {
+          visit(ancestors, timeLapse, (MessageEnd) end, result);
+        }
+
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Compute Semantic candidate Structure for MessageEnd.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<SemanticCandidateContext>
+   * @param end
+   *          MessageEnd
+   */
+  private void visit(Deque<CapellaElement> ancestors, TimeLapse timeLapse, MessageEnd end,
+      List<SemanticCandidateContext> result) {
+
+    // Execution start / SyncCall and ASyncCall
+    if (timeLapse instanceof Execution && ((Execution) timeLapse).getStart() == end) {
+      result.add(new SemanticCandidateContext(ancestors.peek(), timeLapse, true, ancestors.size() + 1));
+      ancestors.push(timeLapse);
+    }
+
+    // Message - SemanticCandidateContext to directly know which execution/instanceRole DNode must be source/target of
+    // the message's DEdge
+    // Handle all cases :
+    // - sending end of the main branch (no execution found in interactionFragmentToTimeLapseCache)
+    // - receiving end of the main branch (start of execution found in interactionFragmentToTimeLapseCache cache and pushed on the ancestor
+    // stack)
+    // - sending end of the return branch (end of execution found in interactionFragmentToTimeLapseCache cache, not yet
+    // removed from the ancestor cache)
+    // - receiving end of the return branch (no execution found in interactionFragmentToTimeLapseCache cache).
+    SequenceMessage message = end.getMessage();
+    if (message != null) {
+      result.add(new SemanticCandidateContext(ancestors.peek(), message, end.equals(message.getSendingEnd()),
+          ancestors.size()));
+    }
+
+    // Execution with return branch end / SyncCall
+    if (timeLapse instanceof Execution && ((Execution) timeLapse).getFinish() == end) {
+      ancestors.pop();
+      result.add(new SemanticCandidateContext(ancestors.peek(), timeLapse, false, ancestors.size() + 1));
+    }
+
+    // @formatter:off
+    // The Following diagram would result in
+    //
+    // | IR1 | | IR2 ] | IR3 ]
+    // | | |
+    // | | |
+    // |-------> - | e1 m1
+    // | | |------> - e2 m2
+    // | | | | |
+    // | | - <-----| | e3 m3
+    // | || | | |
+    // | || | | |
+    // | | - ----->| | return_m3
+    // | | | | |
+    // |<------- - | | return_m1
+    // | | -
+    // | | |
+
```

Comment:
The formatting of this section seems to be off. I think it was messed-up during the save + auto format.

Can you restore it please?

Reply:
Hi Sandu,
Nathalie is on holidays until Monday. 
She will fix this on Monday. 
We need to disable the Eclipse formatter  for this section as it is the only solution to have a understandable view of the computation here.

Reply:
Done!